### ISSUE TITLE
requiring type hint and return type

### ISSUE THREAD

**Issue by mathroc** [squizlabs/PHP_CodeSniffer#2063]
Hi,

before php 7 I used to have those rules:

```xml
    <rule ref="Squiz.Commenting.FunctionComment.Missing"/>
    <rule ref="Squiz.Commenting.FunctionComment.MissingParamTag"/>
    <rule ref="Squiz.Commenting.FunctionComment.MissingParamName"/>
    <rule ref="Squiz.Commenting.FunctionComment.ParamNameNoMatch"/>
    <rule ref="Squiz.Commenting.FunctionComment.IncorrectTypeHint"/>
    <rule ref="Squiz.Commenting.FunctionComment.MissingReturn"/>
```

to make sure that every function had its parameters and return type specified somewhere. now with php 7 I'd like to do specify the types directly without having to repeat that, eg:

```php
// before:
/**
 * @param string $bar
 * @param Option $o
 * @param int[] $ids
 * @return string
 */
function foo($bar, Option $o, array $ids) {
    // …
}

// after:
/**
 * @param int[] $ids
 */
function foo(string $bar, Option $o, array $ids): string {
    // …
}
```

So that's probably multiple sniffs, note that those should fail:

```php
function foo($bar): string {
    // $bar should be type hinted
}
```

```php
/**
 * @param int $bar
 */
function foo(string $bar): string {
    // $bar: string does not include int
}
```

```php
/**
 * @param int[] $ids
 */
function foo($ids): string {
    // $ids should be `array` type hinted
}
```

```php
/**
 * @return string[]
 */
function foo() {
    // function return type should be type hinted `array`
}
```

```php
function foo() {
    // function return type should be type hinted
}
```

for that I propose those sniffs:

```xml
    <rule ref="Generic.Function.MissingParamTypeHint"/>
    <rule ref="Generic.Function.IncorrectParamTypeHint"/>
    <rule ref="Generic.Function.MissingReturnTypeHint"/>
    <rule ref="Generic.Function.IncorrectReturnTypeHint"/>
```

what do you think, would that be useful ?

**Comment by gsherwood** [comment]
A sniff to ensure that function params and return values are type hinted sounds good, although it couldn't tell you if the type hint is accurate.

Another sniff (or a change to the `Squiz.Commenting.FunctionComment` sniff) could be useful for ensuring your documentation (if any) matches what the code says - probably using the code as the authoritative source.

So yes, I think some new sniffs could be useful here.

**Comment by mathroc** [comment]
great

> A sniff to ensure that function params and return values are type hinted sounds good, although it couldn't tell you if the type hint is accurate.

With `IncorrectParamTypeHint` I don't expect phpcs to look if I actually return something valid. I was only thinking of accuracy regarding the comment type hint. but maybe it's best to leave it in `Squiz.Commenting.FunctionComment`

**Comment by mathroc** [comment]
this projects provides smilar sniffs for phpcs : https://github.com/slevomat/coding-standard

**Comment by CZechBoY** [comment]
For typehints for function parameters it is better to use more complex solution like phpstan (https://github.com/phpstan/phpstan-strict-rules). Problem of phpcs is that it looks only on 1 one/class at once. You get errors/warning if parent class has no typehints, if you add typehint you get php error.

**Comment by TomasVotruba** [comment]
To even automate migration from docblocks to typehint you can use Rector, that changes the code and check for conflicts with parent classes (in and out the `/vendor): 
https://www.tomasvotruba.cz/blog/2018/12/10/rocket-science-behind-migration-of-docblock-types-to-php-typehints/#give-your-code-a-typehint-facelift

**Comment by lifesboy** [comment]
> tly without having to repeat

I get this error:
`ERROR: Referenced sniff "Generic.Function.MissingParamTypeHint" does not exist`

**Comment by BeycanDeveloper** [comment]
Yes, i want same configuration but no way

**Comment by ldebrouwer** [comment]
@BeycanDeveloper Please take a moment to read; https://github.com/squizlabs/PHP_CodeSniffer/issues/3932

**Comment by kkmuffme** [comment]
Since all static analysis tools (phpstan, psalm) check for the type and report an error if it's missing (neither phpdoc nor type hint exist), implementing this doesn't really make sense anymore.

Probably the existing phpdoc only errors can be deprecated eventually

*I know this repo is not in use anymore, but opening a new issue in the new repo for this comment didn't make sense

**Comment by rodrigoprimo** [comment]
Looking at this issue, it seems like most of what was requested is now covered:

- **Missing type hints** (`Generic.Function.MissingParamTypeHint` / `MissingReturnTypeHint`): A similar request was opened in the new repo (PHPCSStandards/PHP_CodeSniffer#291) and closed as not planned, as this is already covered by the [Slevomat Coding Standard](https://github.com/slevomat/coding-standard/).
- **Incorrect param type hint** (`Generic.Function.IncorrectParamTypeHint`): Already handled by `Squiz.Commenting.FunctionComment.IncorrectTypeHint`, which flags mismatches between `@param` docblock types and native type hints.
- **Incorrect return type hint** (`Generic.Function.IncorrectReturnTypeHint`): This is the only part not currently covered by a PHPCS sniff, as far as I know.

Given the above, should this issue be closed or kept open because of the suggestion for a `Generic.Function.IncorrectReturnTypeHint` sniff?

**Comment by mathroc** [comment]
I think between the new type hints and what static analysis tools are doing, this is now well covered, thx @rodrigoprimo for the reminder, I'll close this. if someone has something missing, I guess a new issue is welcome on the new repo
